File: src/Services/api_carga_usuario.py
import pandas as pd
import requests

# Función para leer el archivo Excel y preparar los datos de usuarios
def leer_usuarios_excel(archivo_excel):
    # Leer el archivo Excel y asegurar que la columna 'password' sea leída como texto
    df = pd.read_excel(archivo_excel, dtype={'password': str})

    usuarios = []
    for index, row in df.iterrows():
        # Validar y manejar valores faltantes o incorrectos
        password = str(row['password']) if pd.notna(row['password']) else "default_password"  # Usar "default_password" si no hay valor
        saldo = float(row['saldo']) if pd.notna(row['saldo']) else 0  # Convertir el saldo a float, 0 si está vacío

        # Convertir id_entidad a número si es posible, y manejar valores NaN
        try:
            id_entidad = int(float(row['id_entidad'])) if pd.notna(row['id_entidad']) else None
        except ValueError:
            id_entidad = None  # Si no puede convertirse a número, se maneja como None

        # Asegurarse que el tipo de usuario es 'Usuario' o 'Miembro', de lo contrario usar 'Usuario'
        tipo_usuario = row['tipo_usuario'] if row['tipo_usuario'] in ['Usuario', 'Miembro'] else 'Usuario'

        # Crear un diccionario con los datos del usuario
        usuario = {
            'nombre_apellido': row['nombre_apellido'],  # No separar el nombre y apellido
            'dni': str(row['dni']),  # Asegurarse de que el DNI sea una cadena
            'password': password,  # Usar el valor leído o el valor predeterminado
            'tipo_usuario': tipo_usuario,  # Asegurarse que sea 'Usuario' o 'Miembro'
            'id_entidad': id_entidad,  # Dejar el id_entidad como None si es un usuario sin entidad
            'saldo': saldo  # Asignar el saldo leído o 0 si está vacío
        }
        usuarios.append(usuario)  # Añadir cada usuario a la lista de usuarios

    return usuarios  # Devolver la lista de usuarios

# Función para enviar los datos de cada usuario a la API PHP
def enviar_usuarios_api(usuarios):
    url_php = "http://localhost/teodepagos/src/Services/api_carga_usuario.php"  # Cambiar la URL según tu configuración
    headers = {'Content-Type': 'application/json; charset=utf-8'}  # Definir los encabezados para la solicitud POST

    for usuario in usuarios:
        # Imprimir el usuario que estamos enviando para verificar que los datos sean correctos
        print(f"Enviando usuario: {usuario}")
        try:
            # Hacer la solicitud POST a la API PHP
            response = requests.post(url_php, json=usuario, headers=headers)

            # Comprobar la respuesta del servidor
            if response.status_code == 200:
                print(f"Usuario {usuario['nombre_apellido']} enviado correctamente. Respuesta: {response.text}")
            else:
                print(f"Error al enviar el usuario {usuario['nombre_apellido']}. Respuesta: {response.text}")
        except requests.exceptions.RequestException as e:
            # Manejar errores de conexión o solicitudes fallidas
            print(f"Error al conectar con la API: {e}")

if __name__ == '__main__':
    archivo_excel = 'usuarios.xlsx'  # Cambiar la ruta al archivo Excel según tu entorno
    usuarios = leer_usuarios_excel(archivo_excel)  # Leer los usuarios desde el archivo Excel
    enviar_usuarios_api(usuarios)  # Enviar los usuarios a la API

File: src/Services/api_carga_usuario.php
<?php
// Habilitar la visualización de errores
ini_set('display_errors', 1);
ini_set('display_startup_errors', 1);
error_reporting(E_ALL);

// Incluir archivo de configuración
$config = require '../../config/config.php';

// Incluir la clase de la base de datos
require '../../src/Models/Database.php';

// Instanciar la conexión a la base de datos
try {
    $db = new Database($config['db_host'], $config['db_name'], $config['db_user'], $config['db_pass']);
    $pdo = $db->getConnection();
} catch (PDOException $e) {
    die("Error en la conexión a la base de datos: " . $e->getMessage());
}

// Verificar si los datos fueron enviados con el método POST
if ($_SERVER['REQUEST_METHOD'] == 'POST') {
    // Leer el contenido JSON del cuerpo de la solicitud
    $json_data = file_get_contents('php://input');
    $data = json_decode($json_data, true);

    // Verificar si los datos fueron decodificados correctamente
    if ($data === null) {
        echo "Error al decodificar los datos JSON.";
        exit();
    }

    // Verificar si todos los campos necesarios están presentes
    $campos_necesarios = ['nombre_apellido', 'dni', 'password', 'tipo_usuario'];
    $campos_faltantes = [];

    foreach ($campos_necesarios as $campo) {
        if (!isset($data[$campo]) || empty(trim($data[$campo]))) {
            $campos_faltantes[] = $campo;
        }
    }

    if (count($campos_faltantes) > 0) {
        echo "Faltan los siguientes datos: " . implode(', ', $campos_faltantes);
    } else {
        $nombre_apellido = trim($data['nombre_apellido']);
        $dni = $data['dni'];
        $password = isset($data['password']) ? $data['password'] : "default_password";  // Asignar contraseña por defecto si no existe
        $tipo_usuario = $data['tipo_usuario'];
        $saldo = isset($data['saldo']) ? $data['saldo'] : 0;  // Asignar saldo por defecto si no existe
        $id_entidad = isset($data['id_entidad']) ? $data['id_entidad'] : null;  // Solo será usado si es 'Miembro'

        // Verificar si el saldo es numérico
        if (!is_numeric($saldo)) {
            echo "El saldo debe ser un número válido.";
        } else {
            // Crear la consulta SQL para insertar el usuario
            $sql = "INSERT INTO usuarios (nombre_apellido, dni, password, tipo_usuario, id_entidad, saldo) 
                    VALUES (:nombre_apellido, :dni, :password, :tipo_usuario, :id_entidad, :saldo)";

            // Preparar la consulta con PDO
            $stmt = $pdo->prepare($sql);
            
            // Verificar si es un Usuario o un Miembro
            if ($tipo_usuario === 'Usuario') {
                $id_entidad = null;  // Asegurar que id_entidad sea NULL si el tipo es 'Usuario'
            }

            // Ejecutar la consulta
            try {
                $stmt->execute([
                    ':nombre_apellido' => $nombre_apellido,
                    ':dni' => $dni,
                    ':password' => $password,  // Asegúrate de cifrar las contraseñas en producción
                    ':tipo_usuario' => $tipo_usuario,
                    ':id_entidad' => $id_entidad,
                    ':saldo' => $saldo
                ]);

                echo "Usuario registrado correctamente.";

            } catch (PDOException $e) {
                echo "Error al insertar el usuario: " . $e->getMessage();
            }
        }
    }
} else {
    echo "Solo se aceptan solicitudes POST.";
}
?>
